### PR TITLE
Upstream PHP SSL Meterpreter and Shells

### PR DESCRIPTION
**Pull request by sempervictus**
This is old SVIT code, it was likely PRed years back as part of
my efforts to SSL/TLS wrap all command and interpreted language
transports (never thought it appropriate for pentesters to skyline
client data during an engagement). PHP has some peculiarities in
how select() is handled over socket abstractions in userspace, aka,
TLS sockets, and disambiguating them from the Framework-side error
was problematic as introspection of established TLS streams is
(by design) problematic.

Recent use of the included Meterpreter payload showed it to work
without losing session state as it did in years past, with proper
port forwarding and core functionality. While it is significantly
less evolved than the Python version in terms of extensions/API
compatibility with Meterpreter proper or Mettle, the runtime is
far more commonly exposed as a vector for ingress.

**Testing:**
  Executed from multiple systems running PHP5 across local and WAN
links. Source hosts were all Linux, reasonably current revisions of
OS and runtimes/libraries. Port forwarding, file ops, command exec,
and other basic functions appear to work properly.

**TODO:**
  Refactor this PR like a decent human being into one for Payloads
containing the actual PHP SSL meterpreter file, and one for MSF
containing the rest of it.
  Get wider testing - this is most likely not upstream today due
to the unpleasantness associated with different versions of PHP and
SSL socket handling.
  Update php_hop with TLS capability to provide encrypted chains of
transport without full sessions on intermediaries.
  Implement certificate validation to provide authenticated and
encrypted sessions. Port to php_hop to complete the chain.

Ping @OJ, @egypt, @bcook-r7, @scriptjunkie for review (thanks guys).

### CODE REVIEW

File: modules/payloads/singles/php/meterpreter_reverse_tcp_ssl.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      'Description'   => 'Connect back to attacker with SSL and spawn a Meterpreter server (PHP)',
+      'Author'        => ['RageLtMan <rageltman[at]sempervictus>'],
+      'Platform'      => 'php',
+      'Arch'          => ARCH_PHP,
+      'Handler'       => Msf::Handler::ReverseTcpSsl,
+      'License'       => MSF_LICENSE,
+      'Session'       => Msf::Sessions::Meterpreter_Php_Php))
+  end
+
+  def generate
```
Comment (busterb) [inline modules/payloads/singles/php/meterpreter_reverse_tcp_ssl.rb]: I'd like to see some merging of shared code and generate in meterpreter_reverse_tcp.rb.

File: modules/payloads/stagers/php/reverse_tcp_ssl.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      'Arch'          => ARCH_PHP,
+      'Handler'       => Msf::Handler::ReverseTcpSsl,
+      'Stager'        => {'Payload' => ""}
+      ))
+  end
+
+  #
+  # Constructs the payload
+  #
+  def generate
```
Comment (busterb) [inline modules/payloads/stagers/php/reverse_tcp_ssl.rb]: this is so much the same with the meterpreter single above, I'd like to see it merged somehow

**Review by busterb — COMMENTED** [review]
Thanks for the submission! I don't think we can really move forward with the PR as-is, since it has some fundamental things that need fixing, as you noted. But it should be possible to cherry pick out the ideas and make a new one that does fit the current patterns.

**Comment by sempervictus** [comment]
can we reopen this for next week and assign me?

**Comment by busterb** [comment]
Well, let's call this an open issue, but a fresh PR (or set of PRs) is probably in order.

**Comment by sempervictus** [comment]
@busterb: could you peek at this again?

File: lib/msf/base/sessions/meterpreter_php_ssl.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#
+# This class creates a platform-specific meterpreter session type
+#
+###
+class Meterpreter_Php_Ssl < Msf::Sessions::Meterpreter_Php_Php
+  def supports_ssl?
+    true
+  end
+  
+  def initialize(rstream, opts={})
```
[the remark this answers is not in the document]
Comment (sempervictus) [reply lib/msf/base/sessions/meterpreter_php_ssl.rb]: Looks like copy pasta, thanks.

**Comment by sempervictus** [comment]
Need? Not if the concern is the post session configuration comms (after ctlv is figures itself out). Useful though - I do all my interpreted language payloads over TLS if possible. With the cert gen on our side fixed, IDS only complains about it being self signed, bit gets no purview into the meterpreter content being sent prior to l7 encryption state being established.

**Comment by jmartin-r7** [comment]
Testing this with https://github.com/rapid7/metasploit-payloads/pull/234 payloads cannot be successfully generated.  The stageless is missing references to be able to generate a `UUID` and the stages payload is missing `reverse_tcp.php`.  Also many of the comments/requests here still need to be addressed.

```
$ ./msfvenom -p php/meterpreter_reverse_tcp_ssl LHOST=vmnet8 LPORT=4444 > metPhpSsl_vmnet8_4444.php
[-] No platform was selected, choosing Msf::Module::Platform::PHP from the payload
[-] No arch selected, selecting arch: php from the payload
Error: undefined local variable or method `generate_payload_uuid' for #<#<Class:0x00007f89c957c1f0>:0x00007f89c24975e8>
$ ./msfvenom -p php/meterpreter/reverse_tcp_ssl LHOST=vmnet8 LPORT=4444 > metPhpSsl_vmnet8_4444.php
[-] No platform was selected, choosing Msf::Module::Platform::PHP from the payload
[-] No arch selected, selecting arch: php from the payload
Error: No such file or directory @ rb_sysopen - <PATH_TO_FRAMEWORK>/metasploit-framework/data/php/reverse_tcp.php
```

There are also lost of code reuse comments to deal with here.

**Comment by sempervictus** [comment]
@jmartin-r7: thanks for tracking on this. I made a couple of quick updates which _should_ address the generation issues. Untested, but shouldn't eat any pets or house plants... Could you prod at this again when you have some bandwidth? I'll test in the near future as well if you dont get to it.

**Comment by jmartin-r7** [comment]
Sadness :-(
```
[*] Running msftidy.rb in ./.git/hooks/post-merge mode
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/payloads/singles/php/meterpreter_reverse_tcp_ssl.rb:44 - [WARNING] Spaces at EOL
modules/payloads/stagers/php/reverse_tcp_ssl.rb:12 - [WARNING] Spaces at EOL
```

**Comment by jmartin-r7** [comment]
Generation still fails for stageless payload, the staged payload no longer complains during generations however it does not look to send any network traffic when launched locally.

```$ ./msfvenom -p php/meterpreter_reverse_tcp_ssl LHOST=vmnet8 LPORT=4444 > metPhpSsl_vmnet8_4444.php
[-] No platform was selected, choosing Msf::Module::Platform::PHP from the payload
[-] No arch selected, selecting arch: php from the payload
Error: undefined local variable or method `generate_payload_uuid' for #<#<Class:0x00007ffdafb019e0>:0x00007ffdae3fe9f0>```

**Comment by busterb** [comment]
It's been a while, I'm going to close this. Feel free to start a new PR when you're ready. Thanks!

**Pull request closed without merging**